src/Microsoft.OData.Client/Metadata/ClientTypeUtil.cs

```diff
@@ -687,6 +698,72 @@ internal static string GetEnumValuesString(string enumString, Type enumType)
             return string.Join(",", memberValues);
         }
 
+        /// <summary>
+        /// Returns true if the <paramref name="instance"/> is an instance of an open type in the <paramref name="model"/>.
+        /// </summary>
+        /// <param name="instance">The instance to examine</param>
+        /// <param name="model">The client model.</param>
+        /// <returns>Returns true if the <paramref name="instance"/> is an instance of an open type in the <paramref name="model"/></returns>
+        internal static bool IsInstanceOfOpenType(object instance, ClientEdmModel model)
+        {
+            Debug.Assert(instance != null, "instance !=null");
+            Debug.Assert(model != null, "model !=null");
+
+            Type clientType = instance.GetType();
+            ClientTypeAnnotation clientTypeAnnotation = model.GetClientTypeAnnotation(clientType);
+
+            Debug.Assert(clientTypeAnnotation != null, "clientTypeAnnotation != null");
+
+            // TODO: Populate dynamic properties only if this is an open (entity or complex) type
+            return clientTypeAnnotation.EdmType.IsOpen();
+        }
+
+        /// <summary>
+        /// Returns true if the <paramref name="instance"/> contains an non-null dictionary property of string and object
+        /// The dictionary should also not be decorated with IgnoreClientPropertyAttribute
+        /// </summary>
+        /// <param name="instance">Object with expected dictionary property</param>
+        /// <param name="dynamicPropertiesDictionary">Reference to the dictionary</param>
+        /// <returns>true if expected dictionary is found</returns>
+        internal static bool TryGetDynamicPropertiesDictionary(object instance, out IDictionary<string, object> dynamicPropertiesDictionary)
+        {
+            Debug.Assert(instance != null, "instance != null");
+
+            dynamicPropertiesDictionary = default(IDictionary<string, object>);
+
+            PropertyInfo propertyInfo = instance.GetType().GetPublicProperties(true /* instanceOnly */).Where(p =>
+                                p.GetCustomAttributes(typeof(ContainerPropertyAttribute), true).Any() &&
+                                typeof(IDictionary<string, object>).IsAssignableFrom(p.PropertyType)).FirstOrDefault();

```

Comment:
What happens if indeed the object instance has properties of type  IDictionary<string,object> which are not intended to be dynamic properties? In this case it gets .FirstOrDefault(). what happens in such a scenario?

Reply:
@paulodero Note that beyond looking for a property of type IDictionary<string, object> or that implements that interface, we also check that its decorated with an attribute `Microsoft.OData.Client.ContainerProperty`. Adding that attribute is signaling that the dictionary property is the dynamic properties container.

Reply:
I get it. Thanks.